### DISCUSSION TITLE
Quarto rendering pdf, running xelatex 10 times

### DISCUSSION THREAD

**Discussion by jpatteet** [quarto-dev/quarto-cli#3517] [Q&A]
I asked a question on stackoverflow and got an answer that resolved the problem. I am posting both here, maybe it is worth some modifications in quarto or not, let me know what you think.

**Question:**

I am wondering if it is possible to render quarto document faster.

When I render my document I get these messages in the "Background Jobs" window:

```
running xelatex - 1
  This is XeTeX, Version 3.141592653-2.6-0.999994 (TeX Live 2022) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode
  
running xelatex - 2
  This is XeTeX, Version 3.141592653-2.6-0.999994 (TeX Live 2022) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode
  
running xelatex - 3
  This is XeTeX, Version 3.141592653-2.6-0.999994 (TeX Live 2022) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode
```

All the way until:

```
running xelatex - 10
  This is XeTeX, Version 3.141592653-2.6-0.999994 (TeX Live 2022) (preloaded format=xelatex)
   restricted \write18 enabled.
  entering extended mode
  
WARNING: maximum number of runs (9) reached
```

Output created:

Does anyone know what those 10 "running xelatex mean ? Is there a way to decrease this number of runs and what is the impact on the final rendering ?

**Answer:**

Quarto generates PDF via LaTeX by default. The LaTeX processors sometimes require multiple runs, as they step through the document sequentially and need a another run to, for example, update the table of contents after seeing another heading. Usually two runs are enough, but some LaTeX packages require more runs to get the correct output. Ten runs is quite unusual though, there might be something wrong elsewhere.

Quarto uses a heuristic to determine if it needs to run the LaTeX engine once more. Quarto also installs any missing LaTeX packages automatically. However, the tool latexmk is probably better in determining when to do additional calls to xelatex. You can use this tool with Quarto:

```
---
format:
  pdf:
    pdf-engine: latexmk
    pdf-engine-opt: -xelatex
---
```

If the document still requires 10 runs, then there might be an issue with a LaTeX package somewhere.

**Comment:**

I added pdf-engine: latexmk and pdf-engine-opt: -xelatex and now it does not even tell me running xelatex and renders the document very quickly

**Comment by dragonstyle** [comment]
Thank you for sharing this! Could you share the qmd that causes this? I think it it probably is a bug and I'd love to fix it...

**Reply by jpatteet** [reply]
Unfortunately I can't at the moment as it is my PhD and it is yet to be published. 

However, after a few weeks, I have tried rendering it again and I get an error message saying

"PDF file is not valid or corrupted. More information
PDF.js v2.8.335 (identifiant de compilation : 228adbf67)
Message : Invalid PDF structure."

It seems the error now comes from the two code lines I added before 

```
---
format:
  pdf:
    pdf-engine: latexmk
    pdf-engine-opt: -xelatex
---
```

If I remove them it works but as slowly as before and if I create another qmd file with just this yaml I get this same error.

Has there been any updates that could create this? Any idea how I could get rid of this and at the same time find a new solution to my slow rendering problem ?

**Reply by dragonstyle** [reply]
What version of Quarto are you using? We definitely changed the behavior of Quarto with the `latexmk` PDF engine in our 1.3 pre-release, so it may address the issue that you're seeing...

**Reply by jpatteet** [reply]
![quartoversion](https://user-images.githubusercontent.com/81798269/215789426-2163ce40-8daa-4121-84aa-3f8feb6681cd.JPG)
I guess version 1.2.313

**Reply by dragonstyle** [reply]
Ok, I do think 1.3 will have a behavior change, though I don't know whether it will fix exactly what you're seeing. We changed two things that are likely at least related:

1) We support remote and SVG images properly when using non latexmk pdf engines directly.

2) We are properly executing all the Quarto code when using the latexmk pdf engine.

**Reply by jpatteet** [reply]
I have done the update to 1.3 and I do not get the error anymore. The rendering is not as fast as a few months ago but faster than when I do not use latexmk and xelatex, but I don't know if it is normal or not (1min10 for 40pages with lots of images)

**Reply by dragonstyle** [reply]
OK this is good news. Previously, when using the `latexmk` engine with Quarto, a lot of the Quarto functionality itself was being disabled (not intentionally, but due to a bug). I have a feeling that enabling that functionality likely accounts for the performance different (even if you're not using those features, we are still doing things like running filters over the document and post processing that LaTeX where we wouldn't have been previously).

**Reply by cscheid** [reply]
@jpatteet When your thesis is ready to be publicly available, would you consider making the sources publicly available as well? We're interested in tracking performance issues in Quarto and would love to know if there are places where we can easily improve.

**Reply by jpatteet** [reply]
Yes when I am done with my thesis, everything will be publicaly availabe online.
As I am still not familiar with Rstudio and Quarto and commonly struggle with various aspects such as formatting. Do you know where is the best place to ask questions about this ?

**Reply by cscheid** [reply]
We have a good bit of documentation on this. https://quarto.org/docs/get-started/authoring/rstudio.html is a good entry point to using RStudio, then the "Authoring" section in https://quarto.org/docs/guide/

If those don't address your issues, asking in the discussion forums here is the next step.

**Comment by mcanouil** [comment]
Closing this as resolved and in favour of https://github.com/quarto-dev/quarto-cli/issues/3518
